### PR TITLE
fishPlugins: create scope and add some plugins

### PR DESCRIPTION
**Pull request by pacien**
###### Motivation for this change

This bootstraps the fishPlugins scope and build tools for packaging plugins for the fish shell.

Two plugin packages are added:
* [fishtape](https://github.com/jorgebucaran/fishtape): a TAP-based test runner for Fish, often used to test fish plugins,
* [pure](https://github.com/rafaelrinaldi/pure): a pretty, minimal and fast Fish prompt, ported from zsh

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Poking people

CC'ing the most recent committers on the fish module and package as they might be interested in this PR:
* @cole-h
* @gebner
* @adisbladis
* @bouk
* @xaverdh
* @rnhmjoj
* @hyperfekt

### CODE REVIEW

File: pkgs/top-level/all-packages.nix
Comment (rnhmjoj) [inline pkgs/top-level/all-packages.nix:8749]: Can you move `foreign-env` into fish-plugins? It has received a special treatment so far, because it's needed to source the NixOS envronment.

Comment (bouk) [reply pkgs/top-level/all-packages.nix:8749]: Kind of tangental, I created https://github.com/bouk/babelfish to avoid having fish-foreign-env in my shell, because it's super slow in my experience. Spawning a prompt takes more than a second. Might be worth integrating in the future.

And I agree we should move ffe into the fishPlugins scope

Comment (pacien) [reply pkgs/top-level/all-packages.nix]: Package moved and module updated :+1:

File: pkgs/top-level/all-packages.nix
Comment (rnhmjoj) [inline pkgs/top-level/all-packages.nix:8749]: You should define an alias in `pkgs/top-level/aliases.nix`, otherwise this will break existing setups.

Comment (pacien) [reply pkgs/top-level/all-packages.nix]: Amended :+1:

File: pkgs/top-level/aliases.nix
Comment (rnhmjoj) [inline pkgs/top-level/aliases.nix:159]: Also add a date, please.

Comment (pacien) [reply pkgs/top-level/aliases.nix]: Added and amended.

**Comment by rnhmjoj** [comment]
It would be good to add documentation on how to add/use these plugins.
I can't figure how to use them: either with adding `fishPlugins.something` to `environment.systemPackages` or a nix-shell fish doesn't seems to find them, even with `programs.fish.vendor.config.enable = true`.

**Comment by pacien** [comment]
That's strange. This should be enough for the plugin to be detected by fish.

I'm using fish enabled through home-manager's `programs.fish.enable` and adding the packages to `home.packages` is enough for them to be detected.

**Comment by rnhmjoj** [comment]
> I'm using fish enabled through home-manager's programs.fish.enable and adding the packages to home.packages is enough for them to be detected.

Uhm, I don't use home-manager, so I tried:
```
nix-shell -I nixpkgs=/path/to/nixpkgs --pure -p fish fishPlugins.fish-foreign-env -- run fish
```
but `fenv` is not available. Similarly, when I try a NixOS VM with:
```
nix-build -E '(import ./nixos { 
  configuration = { pkgs,... }: { 
    users.users.root.password = "root";
    programs.fish.enable = true;
    environment.systemPackages = with pkgs; [ fishPlugins.pure fishPlugins.fish-foreign-env ];
  }; }).vm'
```

**Comment by pacien** [comment]
It seems like the `fish-foreign-env` package was installing the scripts outside of the standard `vendor_functions.d` directory.
I modified the package and the module accordingly and everything is correctly loaded now.

This is however a breaking change. Should I remove the package alias?

**Comment by rnhmjoj** [comment]
> This is however a breaking change. Should I remove the package alias?

Uhm, it's likely it was never used outside of the NixOS module.
I'd say to add a note in the release notes but keep the alias.

**Comment by pacien** [comment]
Note added to the release notes and amended.

**Comment by rnhmjoj** [comment]
Thank you. I tested the NixOS module and built the manual: looks all good.

File: pkgs/misc/fish-plugins/build-fish-plugin.nix
Comment (rnhmjoj) [inline pkgs/misc/fish-plugins/build-fish-plugin.nix:11]: What about using this function to define a `fish.withPlugins`? It would be useful for quickly testing plugins and running them inside a nix-shell.

Comment (pacien) [reply pkgs/misc/fish-plugins/build-fish-plugin.nix:11]: Good idea, this would be useful.

I think it can be done by creating some `wrapFish` package which would accept lists of plugins, custom functions, completion scripts, initialisation scripts and environment variables as customisation parameters.

This is a bit involved though and I'd prefer to keep that for a separate pull request.

Comment (rnhmjoj) [reply pkgs/misc/fish-plugins/build-fish-plugin.nix:11]: > This is a bit involved though and I'd prefer to keep that for a separate pull request.

Yeah, I agree.

**Review by rnhmjoj — APPROVED** [review]
I don't have any other suggestions to make, the changes look good to me.
I'd wait for more fish users/maintainers to comment before merging, though.

File: pkgs/top-level/all-packages.nix
Comment (cole-h) [inline pkgs/top-level/all-packages.nix:8747]: I think this should go into a new `shells/fish/plugins` directory rather than `misc/fish-plugins`.

**Review by cole-h — CHANGES_REQUESTED** [review]
One small suggestion below, otherwise LGTM. Thanks for this, @pacien! And thanks @rnhmjoj for all your feedback / review.

File: pkgs/top-level/all-packages.nix
Comment (pacien) [reply pkgs/top-level/all-packages.nix]: Indeed. I moved the directory as suggested and amended.

**Review by cole-h — APPROVED** [review]
Diff LGTM. Thanks!

**Comment by pacien** [comment]
(Rebased to fix merrge conflict.)

**Comment by abathur** [comment]
@Pacien I'm not a fish user so I don't have much insight, but it sounds like this breaking change is sneaking up on some nix-darwin users (https://github.com/LnL7/nix-darwin/issues/269). I would guess it's doing the same for HM users?

**Comment by rnhmjoj** [comment]
Yeah, nix-darwin and probably home-manager need to change the foreign-env path as in this diff:
https://github.com/pacien/nixpkgs/commit/d94921db12d9c9e2d3e9c2e6e25e50f03f789ecb#diff-e643907e997ff11fbccddd7dacf176032a320f26f8313eb4b18f435597dca2b0

**Comment by pacien** [comment]
Should we remove the alias `fish-foreign-env` -> `fishPlugins.foreign-env` so that the evaluation fails instead of having runtime breakage? Those packages aren't compatible due to the directory layout changes, so I'm not sure that it made sense to have that alias in the first place.

**Comment by cole-h** [comment]
Yeah, I agree with that.

**Pull request merged**
